File: models/Furo.ts
import {Tile, TileType} from "./Tile";
import {Decoder, Encoder} from "./types";

export enum FuroType {
    Chi = "Chi",
    Pon = "Pon",
    Minkan = "Minkan",
    Ankan = "Ankan"
}

export class Furo {
    constructor(readonly type: FuroType, readonly tile: Tile) {
    }

    static encode: Encoder<Furo> = (data) => {
        let raw: any = {type: data.type, tile: Tile.encode(data.tile)}
        if (data.type === FuroType.Minkan) {
            raw.type = "Kan"
            raw.ankan = false
        } else if (data.type === FuroType.Ankan) {
            raw.type = "Kan"
            raw.ankan = true
        }
        return raw
    }

    static decode: Decoder<Furo> = (raw) => {
        let type: FuroType
        if (raw.type === "Kan") {
            type = raw.ankan ? FuroType.Ankan : FuroType.Minkan
        } else {
            type = raw.type
        }
        return new Furo(type, Tile.decode(raw.tile))
    }

    static parse(tiles: string | Tile[], ankan?: boolean): Furo | undefined {
        if (typeof tiles === 'string') {
            let ankan_ = ankan
            let tiles_: Tile[] | undefined

            if (tiles.length === 5 && tiles[0] == tiles[3] && tiles[0] == '0' && tiles[1] == tiles[2]) {
                ankan_ = true
                tiles_ = Tile.parseTiles(`${tiles[1]}${tiles[1]}${tiles[1]}${tiles[1]}${tiles[4]}`)
            } else {
                tiles_ = Tile.parseTiles(tiles)
            }

            if (tiles_ === undefined) {
                return undefined
            }

            return Furo.parse(tiles_, ankan_)
        } else {
            if (tiles.length == 3) {
                if (tiles[0] == tiles[1] && tiles[1] == tiles[2]) {
                    return new Furo(FuroType.Pon, tiles[0])
                } else {
                    if (tiles.some(it => it.type === TileType.Z)) {
                        return undefined
                    }

                    const tiles_ = [...tiles]
                    tiles_.sort((a, b) => a.compareTo(b))
                    if (tiles[1].distance(tiles[0]) === 1 && tiles[2].distance(tiles[1]) === 1) {
                        return new Furo(FuroType.Chi, tiles[0])
                    } else {
                        return undefined
                    }
                }
            } else if (tiles.length == 4) {
                if (tiles[0] == tiles[1] && tiles[1] == tiles[2] && tiles[2] == tiles[3]) {
                    return ankan ? new Furo(FuroType.Ankan, tiles[0]) : new Furo(FuroType.Minkan, tiles[0])
                }
            }

            return undefined
        }
    }

}

File: python-lib/mahjong_utils/models/furo.py
from abc import ABC
from typing import List, Sequence, Union

from pydantic.dataclasses import dataclass

from .mentsu import Shuntsu, Kotsu, Mentsu
from .tile import Tile, parse_tiles
from .tile_type import TileType


class Furo(Mentsu, ABC):
    """
    副露
    """

    def __encode__(self) -> dict:
        return dict(type=type(self).__name__, tile=self.tile.__encode__())

    @classmethod
    def __decode__(cls, data: dict) -> "Furo":
        t = Tile.__decode__(data['tile'])
        if data['type'] == 'Chi':
            return Chi(t)
        elif data['type'] == 'Pon':
            return Pon(t)
        elif data['type'] == 'Kan':
            return Kan(t, data['ankan'])
        else:
            raise ValueError("invalid type: " + data['type'])

    @staticmethod
    def parse(t: Union[Sequence[Tile], str], ankan: bool = False) -> "Furo":
        if isinstance(t, str):
            if len(t) == 5 and t[0] == t[3] == '0' and t[1] == t[2]:
                t = t[1] + t[1] + t[1] + t[1] + t[4]
                ankan = True
            t = parse_tiles(t)

        if len(t) == 3:
            if t[0] == t[1] == t[2]:
                return Pon(t[0])
            else:
                if t[0].tile_type == TileType.Z or t[1].tile_type == TileType.Z or t[2].tile_type == TileType.Z:
                    raise ValueError(f"invalid tiles: {t}")
                t = sorted(t)
                if t[1] - t[0] == 1 and t[2] - t[1] == 1:
                    return Chi(t[0])
                else:
                    raise ValueError(f"invalid tiles: {t}")
        elif len(t) == 4:
            if t[0] == t[1] == t[2] == t[3]:
                return Kan(t[0], ankan)

        raise ValueError(f"invalid tiles: {t}")


@dataclass(frozen=True)
class Chi(Shuntsu, Furo):
    """
    吃
    """

    def __repr__(self):
        return f"{self.tile.num}{self.tile.num + 1}{self.tile.num + 2}{self.tile.tile_type.lower()}"


@dataclass(frozen=True)
class Pon(Kotsu, Furo):
    """
    碰
    """

    def __repr__(self):
        return f"{self.tile.num}{self.tile.num}{self.tile.num}{self.tile.tile_type.lower()}"


@dataclass(frozen=True)
class Kan(Kotsu, Furo):
    """
    杠
    """

    ankan: bool

    def __encode__(self) -> dict:
        x = Furo.__encode__(self)
        x["ankan"] = self.ankan
        return x

    def __repr__(self):
        if self.ankan:
            return f"0{self.tile.num}{self.tile.num}0{self.tile.tile_type.lower()}"
        else:
            return f"{self.tile.num}{self.tile.num}{self.tile.num}{self.tile.num}{self.tile.tile_type.lower()}"

    @property
    def tiles(self) -> List[Tile]:
        return [self.tile] * 4
